### PR TITLE
Version 3.3.1 - native improvements

### PR DESCRIPTION
**Pull request by konraddysput**
# Backtrace-Unity 3.3.1 update 

This patch adjusts the approach Backtrace-unity uses to detect low memory situations. Previously Backtrace report low memory warning to Backtrace when the lowMemoryWarning callback occurred. This updated approach allows the system to detect out-of-memory crashes via a native OS library and adjustments to the data collected by Backtrace.

**On iOS** - The system reports the out-of-memory exception when the lowMemoryWarning occurred and when the application unexpectedly exits. On the next application startup, Backtrace will analyze the status file. If the status file is available and the following conditions are met, then the Backtrace-Unity integration sends an OutOfMemory exception to Backtrace. The conditions that are checked are:
* application version is the same,
* system version is the same,
* debugger wasn't connected on the last session,
* out of memory warning occurred,
* out of memory resource is available 

**On Android** - The report generated was extended with attributes to let developers know the following information:
* memory warning occurred,
* last time when memory warning occurred

## Testing strategy

1. Integration/Unit tests
- make sure all of them are passing
2. Manual tests
- on Android (oom enabled)
Run the application that generates a lot of textures to see how the application performs with out of memory exception and try to generate crash with attributes

- on iOS (oom enabled)
validate if the status file is generated on application startup with basic information about the application 
validate if the status file has application version, systemVersion, debug information on the application startup,
validate if application stores low memory warning when the lowMemoryWarning event occurred,
validate if application stores report attributes when the lowMemoryWarning event occurred,
validate if application upload the out of memory report when application version, systemVersion, debugging information are the same,
validate if the application doesn't upload the out of memory report when application version, systemVersion, debugging information are the same,

- on iOS (oom disabled)
check if status json file is generated,
check if notification are generated when lowMemoryWarning or other events happening.

### CODE REVIEW

File: Runtime/BacktraceClient.cs [hunk 1]
```diff
@@ -708,17 +708,17 @@ internal void HandleLowMemory()
...
-                var oom = new BacktraceUnhandledException(lowMemoryMessage, string.Empty);
-                SendUnhandledException(oom);
+                // inform native layer about oom error
+                _nativeClient.OnOOM();
+
             }
+            const string lowMemoryMessage = "OOMException: Out of memory detected.";
+            _backtraceLogManager.Enqueue(new BacktraceUnityMessage(lowMemoryMessage, string.Empty, LogType.Error));
+
+
```
Comment (rqbacktrace) [inline Runtime/BacktraceClient.cs:721]: Extra space

File: Runtime/Model/BacktraceConfiguration.cs [hunk 2]
```diff
@@ -104,15 +104,27 @@ public class BacktraceConfiguration : ScriptableObject
...
+        [Tooltip("(Early access) Send Low memory warnings to Backtrace")]
+#elif UNITY_IOS
+         /// <summary>
+        /// Send Out of memory exceptions to Backtrace. 
+        /// </summary>
+        [Tooltip("(Early access) Send Out of memory exceptions to Backtrace")]
+#endif
+        public bool OomReports = false;
+
+
```
Comment (rqbacktrace) [inline Runtime/Model/BacktraceConfiguration.cs:126]: 2 extra newlines

File: Runtime/Native/iOS/NativeClient.cs [hunk 3]
```diff
@@ -205,12 +205,11 @@ public bool OnOOM()
             {
                 return false;
             }
-            // set temporary attribute to "Hang"
-            SetAttribute("error.type", "Low Memory");
-            NativeReport("OOMException: Out of memory detected.");
-            // update error.type attribute in case when crash happen 
-            SetAttribute("error.type", "Crash");   
-            
+
```
Comment (rqbacktrace) [inline Runtime/Native/iOS/NativeClient.cs:208]: Extra newline

**Comment by rqbacktrace** [comment]
Looks good, just found some spacing issues

File: Editor/BacktraceConfigurationLabels.cs [hunk 4]
```diff
@@ -10,6 +10,13 @@ internal static class BacktraceConfigurationLabels
         internal static string LABEL_DESTROY_CLIENT_ON_SCENE_LOAD = "Destroy client on new scene load (false - Backtrace managed)";
         internal static string LABEL_SAMPLING = "Log random sampling rate";
         internal static string LABEL_HANDLE_ANR = "Handle ANR (Application not responding)";
+
+        internal static string LABEL_HANDLE_OOM =
+#if UNITY_ANDROID
+            "(Early access) Send Low memory warnings to Backtrace";
+#else
+            "(Early access) Send Out of memory exceptions to Backtrace";
```
Comment (rqbacktrace) [inline Editor/BacktraceConfigurationLabels.cs:18]: Maybe we can guard this `else` statement a bit more, unless all other platforms support out of memory exceptions

File: Runtime/Model/JsonData/BacktraceAttributes.cs [hunk 5]
```diff
@@ -66,7 +66,29 @@ public BacktraceAttributes(BacktraceReport report, Dictionary<string, string> cl
             SetMachineAttributes(onlyBuiltInAttributes);
             SetProcessAttributes(onlyBuiltInAttributes);
             SetSceneInformation(onlyBuiltInAttributes);
+#if UNITY_ANDROID
```
Comment (rqbacktrace) [inline Runtime/Model/JsonData/BacktraceAttributes.cs:69]: Maybe we want to change this to a generic `ReadPlatformSpecificAttributes` function. Then we can get rid of this compile flag too.

File: Runtime/BacktraceClient.cs (hunk 1 above)
Comment (konraddysput) [reply Runtime/BacktraceClient.cs:721]: done.

File: Runtime/Model/BacktraceConfiguration.cs (hunk 2 above)
Comment (konraddysput) [reply Runtime/Model/BacktraceConfiguration.cs:126]: done.

File: Runtime/Native/iOS/NativeClient.cs (hunk 3 above)
Comment (konraddysput) [reply Runtime/Native/iOS/NativeClient.cs:208]: done.

File: Editor/BacktraceConfigurationLabels.cs (hunk 4 above)
Comment (konraddysput) [reply Editor/BacktraceConfigurationLabels.cs:18]: done.

File: Runtime/Model/JsonData/BacktraceAttributes.cs (hunk 5 above)
Comment (konraddysput) [reply Runtime/Model/JsonData/BacktraceAttributes.cs:69]: Moving this code to Android NativeClient

**Review by rqbacktrace — APPROVED** [review]

**Review by jasoncdavis0 — APPROVED** [review]
Updated the change log and description of PR.

**Review by vlussenburg — APPROVED** [review]

**Pull request merged**